docs/proposals/20220210-md-node-label-sync.md

```diff
@@ -0,0 +1,159 @@
+- Support for arbitrary/user-specified label prefixes. 
+
+## Proposal
+
+CAPI will define its own label prefix which falls within the [restricted namespace *.k8s.io/](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/279-limit-node-access):
+
+```
+node.cluster.k8s.io/*
+```
+
+Labels placed on the Machine that match the above prefixe will be authoritative -- Cluster API will act as the ultimate authority for anything under the prefix. Labels with the above prefix should be set only within Cluster API CRDs.
+
+In addition to the above, CAPI will support one-time application of the following labels during Machine creation:
+
+```
+node-role.kubernetes.io/*
+node-restriction.kubernetes.io/*
+```
+
+The `node-role.kubernetes.io` prefix has been used widely in the past to identify the role of a node (e.g. `node-role.kubernetes.io/worker=''`). For example, `kubectl get node` looks for this specific label when displaying the role to the user.
+
+The use of the `node-restriction.kubernetes.io/*` prefix is recommended in the [Kubernetes docs](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) for things such as workload placement.
+
+It's important to note that the these specific labels are *not* kept in sync, but rather applied only once after the Machine and its corresponding Node are first created. Taking authoritative ownership of these prefixes would preclude other entities from adding and removing such labels on the Node. 
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin/user, I would like a declarative and secure means by which to assign roles to my nodes. For example, when I do `kubectl get nodes`, I want the ROLES column to list my assigned roles for the nodes.
+
+#### Story 2
+
+As a cluster admin/user, for the purpose of workload placement, I would like a declarative and secure means by which I can add/remove labels on groups of nodes. For example, I want to run my outward facing nginx pods on a specific set of nodes (e.g. md-secure-zone) to reduce exposure to the company's publicly trusted server certificates.
+
+### Implementation Details/Notes/Constraints
+
+#### Label propogation from MachineDeployment to Machine
+
+Option #1: CAPI metadata Propagation
+
+CAPI supports [metadata propagation](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/metadata-propagation.html?highlight=propagation#metadata-propagation) across core API resources. For the MachineDeployment types, template labels propagate to MachineSets top-level and MachineSets template metadata:
+```
+MachineDeployment.spec.template.metadata.labels => MachineSet.labels, MachineSet.spec.template.metadata.labels
+```
+Similarly, on MachineSet, template labels propagate to Machines:
+```
+MachineSet.spec.template.metadata.labels => Machine.labels
+```
+
+As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. 
```

Comment:
I'd suggest to keep this a the first option and update it to propose adding support in the MD to recognise these labels and exclude them from the comparison for upgrades similar to what we do for MachineDeploymentUniqueLabel and move the other options under the alternative sections.

Reply:
@enxebre That's a good idea!
Looking at the current code, it would mean skipping the comparison of these specific prefixes in `EqualMachineTemplate()`. 

Any other gotchas here?

Reply:
The main gotcha with this stuff is that we have to make it backwards compatible (i.e. a rollout of a CAPI version with the new hash calculation should not automatically trigger rollouts).

But not sure if you have to mention that in the proposal.